File: 2721-sum-of-distances/2721-sum-of-distances.cpp
class Solution {
public:
    vector<long long> distance(vector<int>& nums) {
        int n = nums.size();
        vector<long long> res (n,0);

        unordered_map<int,vector<int>> d;
        for (int i = 0; i < n; i++){
            d[nums[i]].push_back(i);
        }

        for (auto it : d){
            int num = it.first;
            vector<int>& indexes = it.second;

            long long totalSum = 0;
            for (int index : indexes){
                totalSum += (long long )index;
            }
            long long preSum = 0;

            long long total_sum = 0;
            for (int i = 0 ; i < indexes.size() ; i++){
                int index = indexes[i];
                long long postSum = totalSum - preSum - index;

                // i - p1 + i- p2 .....
                res[index] += (index* (long long )i); // pre sum all i s
                res[index] -= (preSum);  // substract all the prev indexes

                // a1 - i + a2 -i + ...

                res[index] -= (index* (long long)(indexes.size()-i-1)); // pre sum all i s
                res[index] += (postSum);

                preSum += index;

            }
        }

        return res;
    }
};

File: 2721-sum-of-distances/2721-sum-of-distances.rs
use std::collections::HashMap;
impl Solution {
    pub fn distance(nums: Vec<i32>) -> Vec<i64> {
        let n = nums.len();
        let mut res = vec![0; n];
        let mut d: HashMap<i32, Vec<usize>> = HashMap::new();

        for (i, num) in nums.iter().enumerate() {
            d.entry(*num).or_insert(Vec::new()).push(i);
        }

        for (_, indexes) in d {
            let total_sum: i64 = indexes.iter().map(|&index| index as i64).sum();
            let mut pre_sum: i64 = 0;

            for (i, &index) in indexes.iter().enumerate() {
                let post_sum = total_sum - pre_sum - index as i64;

                // i - p1 + i - p2 .....
                res[index] += (index as i64) * (i as i64); // pre sum all i s
                res[index] -= pre_sum; // subtract all the previous indexes

                // a1 - i + a2 - i + ...
                res[index] -= (index as i64) * ((indexes.len() - i - 1) as i64); // pre sum all i s
                res[index] += post_sum;

                pre_sum += index as i64;
            }
        }

        res
    }
}